pkg/server/extended.go

```diff
@@ -1021,25 +1026,177 @@ func (s *Server) ExtendedCreateThread(w http.ResponseWriter, r *http.Request) {
 				thread.ID,
 			}
 
-			if err = create(s.db.WithContext(r.Context()), new(db.Message), publicMessage); err != nil {
+			if err := create(tx, new(db.Message), publicMessage); err != nil {
 				w.WriteHeader(http.StatusBadRequest)
 				_, _ = w.Write([]byte(err.Error()))
-				return
+				return err
 			}
 		}
+
+		return nil
+	}); err != nil {
+		slog.Error("failed to create thread")
 	}
 
 	writeObjectToResponse(w, thread.ToPublic())
 }
 
-func (s *Server) ExtendedCreateThreadAndRun(w http.ResponseWriter, _ *http.Request) {
-	// TODO: When this does eventually get implemented, we need to emit three events:
-	// 1. The first event is a thread.created event that says that the thread is created.
-	// 2. The second event is a thread.run.created event that says that the run is created.
-	// 3. The third event is a thread.run.queued event that says that the run is queued.
-	// Note that this requires that the EventIndex field on the resulting Run object is set to 2.
-	//TODO implement me
-	w.WriteHeader(http.StatusNotImplemented)
+func (s *Server) ExtendedCreateThreadAndRun(w http.ResponseWriter, r *http.Request) {
+	createThreadAndRunRequest := new(openai.ExtendedCreateThreadAndRunRequest)
+	if err := readObjectFromRequest(r, createThreadAndRunRequest); err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		_, _ = w.Write([]byte(err.Error()))
+		return
+	}
+
+	if err := validateMetadata(createThreadAndRunRequest.Metadata); err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		_, _ = w.Write([]byte(err.Error()))
+		return
+	}
+
+	//nolint:govet
+	publicThread := &openai.ThreadObject{
+		// The first two fields will be set on create.
+		0,
+		"",
+		createThreadAndRunRequest.Metadata,
+		openai.ThreadObjectObjectThread,
+	}
+
+	var (
+		gormDB = s.db.WithContext(r.Context())
+		thread = new(db.Thread)
+	)
+	if err := gormDB.Transaction(func(tx *gorm.DB) error {
+		if err := create(tx, thread, publicThread); err != nil {
+			w.WriteHeader(http.StatusBadRequest)
+			_, _ = w.Write([]byte(err.Error()))
+			return err
+		}
+
+		if publicThread := createThreadAndRunRequest.Thread; publicThread != nil {
```

Comment:
Is it worth doing this?
```suggestion
		if publicThread := createThreadAndRunRequest.Thread; publicThread != nil && publicThread.Messages != nil {
```

Reply:
I initially had this and removed it. Your comment has convinced me. Added.